app/services/course/user_registration_service.rb

```diff
@@ -161,10 +161,26 @@ def invalid_code(registration)
   def accept_invitation(registration, invitation)
     registration.course_user = invitation.course_user
     invitation.course_user.accept!(registration.user)
-    invitation.course_user.save
+    CourseUser.transaction do
+      assign_email_to_user(invitation.user_email, invitation.course_user.user)
+      invitation.course_user.save
+    end
     invitation.course_user
   end
 
+  # Assigns an unclaimed email address to a given user.
+  #
+  # @param [User::Email] email
+  # @param [User] user
+  # @raise [IllegalStateError] when the email already belongs to another user.
+  def assign_email_to_user(email, user)
+    if email.user.nil?
+      email.update(user: user)
+    elsif email.user != user
+      raise IllegalStateError
```

Comment:
I think you should raise ` ActiveRecord::Rollback` ? otherwise the user won't be able to register because `IllegalStateError`